### PR TITLE
[Keyboard] Add Chouchou keyboard

### PR DESCRIPTION
**Pull request by dlip**
## Description

Add [Chouchou](https://github.com/dlip/chouchou) keyboard and [Taipo](https://inkeys.wiki/en/keymaps/taipo) layout implementation

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/chouchou/keymaps/default/keymap.c [hunk 1]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2023 Dane Lipscombe (@dlip)
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include QMK_KEYBOARD_H
+#include "users/dlip/taipo.h"
```
Comment (zvecr) [inline keyboards/chouchou/keymaps/default/keymap.c:4]: default keymaps should not pull in userspace

File: keyboards/chouchou/config.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2023 Dane Lipscombe (@dlip)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_TIMEOUT 1000U
+
+#define DEBOUNCE 20
```
Comment (zvecr) [inline keyboards/chouchou/config.h:9]: Should be defined in info.json

File: keyboards/chouchou/config.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#define ONESHOT_TIMEOUT 500
+
+// 1000hz Polling Rate
+#define USB_POLLING_INTERVAL_MS 1
```
Comment (zvecr) [inline keyboards/chouchou/config.h:13]: Is default value
```suggestion
```

File: users/dlip/taipo.h [hunk 4]
```diff
@@ -0,0 +1,540 @@
+// Copyright 2023 Dane Lipscombe (@dlip)
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Special thanks to Whorf for his help with this
+
+#pragma once
+#include "keycodes.h"
+#include "quantum_keycodes.h"
+#define IDLE_TIMEOUT_MS 150
+#define TAIPO_KEYCODE_OFFSET (SAFE_RANGE + 500)
```
Comment (zvecr) [inline users/dlip/taipo.h:9]: Indexing like this is invalid. As it stands it overflows into the unicode address space.

**Review by zvecr — CHANGES_REQUESTED** [review]

**Comment by dlip** [comment]
@zvecr Thanks for the review, I have updated the code as suggested

File: keyboards/chouchou/README.md
Comment (drashna) [inline keyboards/chouchou/README.md:1]: This file needs to be renamed to be entirely lowercase.  

You can do this from the web interface, if needed.

File: keyboards/chouchou/README.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# Chouchou
+
+![Chouchou](https://i.imgur.com/BYk2n5C.jpeg)
```
Comment (drashna) [inline keyboards/chouchou/README.md:3]:
```suggestion
![Chouchou](https://i.imgur.com/BYk2n5Ch.jpeg)
```

This uses a lower resolution version of the image that fits better into github's markdown rendering.

File: keyboards/chouchou/README.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+- Hold the boot button on your RP2040-Zero while connecting the usb.
+- Copy the file `.build/chouchou_taipo.uf2` to the drive that appears.
+
+## Features
+
+- [Taipo](https://inkeys.wiki/en/keymaps/taipo) (Layer 0). See my userspace [Taipo docs](../../users/dlip/taipo.md) for details
+- WASD gaming (Layer 1/2)
+- Gamepad (Layer 3). See my userspace [Gamepad docs](../../users/dlip/gamepad.md) for details
+
```
Comment (drashna) [inline keyboards/chouchou/README.md:41]: Please stick to the readme template.
```suggestion

See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).

## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Briefly press the button on the back of the PCB - some may have pads you must short instead
* **Keycode in layout**: Press the key mapped to `QK_BOOT` if it is available
```

File: keyboards/chouchou/keymaps/taipo/config.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2023 Dane Lipscombe (@dlip)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_TIMEOUT 1000U
```
Comment (drashna) [inline keyboards/chouchou/keymaps/taipo/config.h:7]: these should be moved to the keyboard's config.h, rather than here.

**Review by drashna — CHANGES_REQUESTED** [review]

File: users/dlip/README.md
Comment (zvecr) [inline users/dlip/README.md:1]: Looks like i lost my previous review comment. But we generally want userspace submissions to be separate to keyboard PRs.

File: users/dlip/gamepad.h [hunk 8]
```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 Dane Lipscombe (@dlip)
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Code forked from https://github.com/brentaro/QMK_HID_Gamepad_with_SOCD
+#pragma once
+#include "keycodes.h"
+#include "quantum_keycodes.h"
+#include "joystick.h"
+
+enum gamepad_keycodes { GP_DPU = SAFE_RANGE, GP_DPD, GP_DPL, GP_DPR, GP_X, GP_A, GP_B, GP_Y, GP_LB, GP_LT, GP_LSB, GP_RB, GP_RT, GP_RSB, GP_STA, GP_BCK, GP_HOM };
```
Comment (zvecr) [inline users/dlip/gamepad.h:9]: You now have 2 blocks of code using `SAFE_RANGE`, which will not work.

**Review by zvecr — CHANGES_REQUESTED** [review]

File: users/dlip/gamepad.h (hunk 8 above)
Comment (dlip) [reply users/dlip/gamepad.h:9]: Is there any better approaches than just adding a number to it?

File: users/dlip/README.md
Comment (dlip) [reply users/dlip/README.md:1]: ok, I'll split it up

**Comment by dlip** [comment]
Thanks again for the review @zvecr and @drashna, I have removed the user space additions and made the suggested updates

File: users/dlip/gamepad.h (hunk 8 above)
Comment (dlip) [reply users/dlip/gamepad.h]: I'll put a `custom_keycodes.h` file in my other pr

**Review by drashna — APPROVED** [review]

File: keyboards/chouchou/info.json [hunk 9]
```diff
@@ -0,0 +1,79 @@
+{
```
Comment (fauxpark) [inline keyboards/chouchou/info.json:1]: Please run `qmk format-json`.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/chouchou/info.json (hunk 9 above)
Comment (dlip) [reply keyboards/chouchou/info.json:1]: done

**Review by fauxpark — APPROVED** [review]

**Pull request merged**